module/zfs/vdev_raidz_math.c

```diff
@@ -0,0 +1,713 @@
+	uint64_t bench_parity;
+	int i, c, fn, impl;
+	int bench_rec_orig[7][3] = {
+		{1, 2, 3}, {0, 2, 3},
+		{0, 1, 3}, {2, 3, 4},
+		{1, 3, 4}, {0, 3, 4},
+		{3, 4, 5}
+	};
+
+	/* init & vdev_raidz_impl_lock */
+	rw_init(&vdev_raidz_impl_lock, NULL, RW_DEFAULT, NULL);
+
+#if !defined(_KERNEL)
+	/*
+	 * It is possible that this routine is called multiple times
+	 * inside user-space test. Check if we already initialized the
+	 * fastest ops and skip the benchmark.
+	 */
+	if (raidz_math_initialized) {
+		goto init_done;
+	}
+#endif
+
+	/* init scalar multiplication lookup table */
+	raidz_init_scalar_mul_lt();
+
+	/* move supported maths into raidz_supp_maths */
+	for (i = 0, c = 0; i < impl_num; i++) {
+		curr_impl = raidz_all_maths[i];
+
+		/* init kstat */
+		init_raidz_kstat(raidz_math_kstats + i, curr_impl->name);
+
+		if (curr_impl->is_supported()) {
+			raidz_supp_maths[c++] = (raidz_math_ops_t *) curr_impl;
+		} else {
+			bench_print(BENCH_SKIP_FMT, curr_impl->name);
+			continue;
+		}
+	}
+	raidz_supp_maths_cnt = c;	/* number of supported algos */
+
+	/* init  kstat for original algo  routines */
+	init_raidz_kstat(&(raidz_math_kstats[impl_num]), "original");
+
+	/* Fake an zio and run the benchmark on it */
+	bench_zio = kmem_alloc(sizeof (zio_t), KM_SLEEP);
+	VERIFY(bench_zio);
+	bzero(bench_zio, sizeof (zio_t));
+
+	bench_zio->io_offset = 0;
+	bench_zio->io_size = BENCH_ZIO_SIZE; /* only data columns */
+	bench_zio->io_data = zio_data_buf_alloc(BENCH_ZIO_SIZE);
+	if (!bench_zio->io_data) {
+		cmn_err(CE_WARN, "RAIDZ: Can not allocate benchmark memory!");
+		/* use last supported math as the fastest */
+		cmn_err(CE_WARN, "RAIDZ: Using [%s] implementation",
+			raidz_supp_maths[raidz_supp_maths_cnt-1]->name);
+		memcpy(&vdev_raidz_fastest_impl,
+			raidz_supp_maths[raidz_supp_maths_cnt-1],
+			sizeof (raidz_math_ops_t));
+
+		kmem_free(bench_zio, sizeof (zio_t));
+		return;
+	}
+
+	/*
+	 * Benchmark new RAIDZ implementations
+	 */
+
+	/* Find the fastest gen functions */
+	for (fn = 0; fn < RAIDZ_GEN_NUM; fn++) {
+		bench_parity = fn+1;
+		best_run_count = 0;
+
+		/* New raidz_map is needed for rach generate_p/q/r */
+		bench_rm = vdev_raidz_map_alloc(bench_zio, 9,
+		    BENCH_D_COLS+bench_parity, bench_parity);
+
+		for (impl = 0; impl < impl_num; impl++) {
+			curr_impl = raidz_all_maths[impl];
+
+			if (!curr_impl->is_supported()) {
+				continue;
+			}
+
+			kpreempt_disable();
+
+			run_count = 0;
+			t_start = gethrtime();
+
+			do {
+				for (i = 0; i < 10; i++, run_count++) {
+					curr_impl->gen[fn](bench_rm);
+				}
+			} while (gethrtime() < (t_start + BENCH_NS));
+
+			speed = run_count * (bench_zio->io_size >> 10);
+			speed /= (BENCH_D_COLS+bench_parity);
+			speed *= 1000000000ULL;
+			speed /= (gethrtime() - t_start);
+
+			kpreempt_enable();
+
+			raidz_math_kstats[impl].gen_kstat[fn].value.ui64 =
+				speed;
+			bench_print(BENCH_RES_FMT, raidz_gen_name[fn],
+				curr_impl->name, speed/1000, speed%1000);
```

Comment:
There a lot of redundant code here it would be good if you could refactor lines 393-414 and 440-458 somehow in to a generic `gen` and `rec` benchmark function.  If possible I'd also recommend adding the benchmark for the original  implementation after the `for (impl = 0; impl < impl_num; impl++) { }` loop.  That way you don't need to second pair of RAIDZ_GEN_NUM/RAIDZ_REC_NUM loops.  I think that would make this code much more readable.